direct/src/gui/DirectFrame.py

```diff
@@ -50,166 +52,113 @@ def __init__(self, parent = None, **kw):
             ('textMayChange',  1,          None),
-            )
+        )
         # Merge keyword options with default options
         self.defineoptions(kw, optiondefs,
-                           dynamicGroups = DirectFrame.DefDynGroups)
+                           dynamicGroups=DirectFrame.DefDynGroups)
 
         # Initialize superclasses
         DirectGuiWidget.__init__(self, parent)
 
         # Call option initialization functions
         self.initialiseoptions(DirectFrame)
 
-    def destroy(self):
-        DirectGuiWidget.destroy(self)
-
     def clearText(self):
         self['text'] = None
         self.setText()
 
     def setText(self, text=None):
-        if text is not None:
-            self['text'] = text
-
-        # Determine if user passed in single string or a sequence
-        if self['text'] == None:
-            textList = (None,) * self['numStates']
-        elif isinstance(self['text'], stringType):
-            # If just passing in a single string, make a tuple out of it
-            textList = (self['text'],) * self['numStates']
-        else:
-            # Otherwise, hope that the user has passed in a tuple/list
-            textList = self['text']
-        # Create/destroy components
-        for i in range(self['numStates']):
-            component = 'text' + repr(i)
-            # If fewer items specified than numStates,
-            # just repeat last item
-            try:
-                text = textList[i]
-            except IndexError:
-                text = textList[-1]
-
-            if self.hascomponent(component):
-                if text == None:
-                    # Destroy component
-                    self.destroycomponent(component)
-                else:
-                    self[component + '_text'] = text
-            else:
-                if text == None:
-                    return
-                else:
-                    from .OnscreenText import OnscreenText
-                    self.createcomponent(
-                        component, (), 'text',
-                        OnscreenText,
-                        (), parent = self.stateNodePath[i],
-                        text = text, scale = 1, mayChange = self['textMayChange'],
-                        sort = DGG.TEXT_SORT_INDEX,
-                        )
+        self.new_set("text", text)
 
     def clearGeom(self):
         self['geom'] = None
         self.setGeom()
 
-    def setGeom(self, geom=None):
-        if geom is not None:
-            self['geom'] = geom
-
-        # Determine argument type
-        geom = self['geom']
-
-        if geom == None:
-            # Passed in None
-            geomList = (None,) * self['numStates']
-        elif isinstance(geom, NodePath) or \
-             isinstance(geom, stringType):
-            # Passed in a single node path, make a tuple out of it
-            geomList = (geom,) * self['numStates']
+    def long_type_check(self, name, object_):
```

Comment:
This should probably be private, and named more descriptively.

More generally, I'm not entirely convinced that having a bunch of if checks for each individual property in here is the best way to do it.  Is there a way to move the type check back to the individual setters, while still keeping common code in `new_set`?  Or to pass the permitted types as an argument to `new_set`, or otherwise breaking up `new_set` into multiple methods?

Reply:
Generally yes, specially here it's not that useful.

The reason long_type_check exists, is because the loop in new_set, always has one of three inputs:

 * a tuple of (None,) with the length of self["numstates"], 
 * a tuple of (object,) copies, with the length of self["numstates"]
 * or it's assumed that whatever was given to the function already is a tuple of (object,) with length of self["numstates"].

We can move that around, but I don't think it would improve stuff.

I will move the type specific variables to the old setters, so it's more obvious which variables come from where and what stays constant during new_set.

As a side node, the tuple of (None,) is only used for destruction of components, so not even that is "cleanly" following the purpose of what a setter should do.

That whole part should be done better anyway. Like, maybe there should be a set_all_states_XXX function that takes the tuple and a set_single_state_XXX that takes the single values.